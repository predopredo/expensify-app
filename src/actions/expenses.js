// NPM MODULES
import uuid from 'uuid'
// FIREBASE DATABASE
import database from '../firebase/firebase';

// NO DB
// component calls action generator
// actions generator returns object
// component dispatches object
// redux store changes

// with DB
// component calls action generator
// actions generator returns function
// component dispatches function (?)
// function runs (has the ability to dispatch other actions and do whatever it wants)

// SET_EXPENSES
export const startSetExpenses = () => {
  return (dispatch) => {              //this function has access to dispatch (redux-thunk)
    // needs to add return so you can chain .then on app.js and render stuff
    return database.ref('expenses').once('value').then((snapshot) => {   //reads once from firebase and then...

      const expenses = [];
      snapshot.forEach((childSnapshot) => {            //populates the empty array
        expenses.push({
          id: childSnapshot.key,                       //sets firebase location (key) to id and saves it on redux store
          ...childSnapshot.val()                       //so we can use it for routing
        });

        dispatch(setExpenses(expenses));               //dispatches the populated array with mapped ids
      });
    });

  };
};

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// ADD_EXPENSE

// ***only works with redux-thunk wired up (on configureStore.js)***
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData; // //created consts: description, note, amount and createdAt that have default values if no value is passed as argument
    const expense = { description, note, amount, createdAt }; // forms an object with above consts

    return database.ref('expenses').push(expense) // return for testing, won't be used in front-end
      .then((ref) => { // ref = newly added to database expense
        dispatch(addExpense({ // when new expense is written on db, you call dispatch and get id and everything else inside an object to store (and trigger rendering on connected components)
          id: ref.key,
          ...expense // you just get the id from the db, the rest is already on the front-end JS
        }));
      });
  };
};

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE', // Action type
  expense
});

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      })
  }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates)
      .then(() => {
        dispatch(editExpense(id, updates));
      })
  }
};