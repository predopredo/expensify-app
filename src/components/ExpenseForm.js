// NPM MODULES
import moment from 'moment';
import React from 'react';
// Airbnb date picker:
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';

// const now = moment();
// console.log(now.format('MMM Do, YYYY - HH:mm'));

export default class ExpenseForm extends React.Component {

  state = {
    description: this.props.expense ? this.props.expense.description : '',
    note: this.props.expense ? this.props.expense.note : '',
    amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
    createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
    calendarFocused: false,
    error: ''
  };

  onDescriptionChange = (event) => {
    const description = event.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = (event) => {
    const note = event.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = (event) => {
    const amount = event.target.value;
    // As it is a controlled input, the value will be always conditioned to the state so if you condition the user input to only get set as state if it matches a specific regex, anything not matching will simply be ignored and won't get into input value. That can be used as validation
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  };

  // airbnb date picker handlers:
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onDateFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  // Form Submission
  onSubmit = (event) => {
    event.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount!' }))
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmitToDispatch({
        description: this.state.description,
        amount: parseFloat(this.state.amount * 100, 10),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      })
    }

  };

  // Rendering JSX
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onDateFocusChange}
            numberOfMonths={1}
            displayFormat="DD/MM/YYYY"
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <button>{this.props.expense ? 'Update Expense' : 'Add Expense'}</button>
        </form>
      </div>
    );
  };
};

