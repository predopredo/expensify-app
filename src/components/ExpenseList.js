// NPM MODULES
import React from 'react';
// react-redux
import { connect } from 'react-redux';
// Components
import ExpenseListItem from './ExpenseListItem';
// Selectors
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
  <div>
    {
      props.expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
          props.expenses.map((expense) => {
            return <ExpenseListItem key={expense.id} {...expense} />
          })
        )
    }
  </div>
);

// When you connect the component to a store, it will automatically re-render when store state changes

const mapStateToProps = (storedState) => { // It automatically re-runs when state changes.
  return {
    expenses: selectExpenses(storedState.expenses, storedState.filters) // applies filters every time filters change
  };
};

export default connect(mapStateToProps)(ExpenseList);
// Connect gets the state of the store provided by the react-redux Provider component (passed down to it's children router which will access components)