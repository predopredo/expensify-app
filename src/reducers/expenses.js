// Default state:
const expensesReducerDefaultState = [];

// Expenses Reducer
export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {

    case 'SET_EXPENSES':
      return action.expenses

    case 'ADD_EXPENSE':
      return [...state, action.expense];

    case 'REMOVE_EXPENSE':
      return state.filter((expense) => expense.id != action.id); // Only keeps items with id != from provided ((action.id) [when expense.id != action.id returns true]

    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates // get matching id expense and add each property from action.updates
          }
        } else {
          return expense; // if id doesn't match, just add the expense w/o modifying it
        }
      })

    default:
      return state;

  }
};
// ps: on REMOVE_EXPENSE: you can use state.filter(({ id }) => id != action.id) so filter({ id }) /=/ filter((expense) => expense.id ...) you can destructure a property of an object and pass it as the filter parameter to use it in filter condition