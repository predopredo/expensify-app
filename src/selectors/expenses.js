// NPM MODULES
import moment from 'moment';
// ***GET VISIBLE EXPENSES***

export default (expenses, { text, sortBy, startDate, endDate }) => {
  
  return expenses.filter((expense) => {

    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(moment(createdAtMoment, 'day')) : true; //moment compare functions
    const endDateMatch = endDate ? endDate.isSameOrAfter(moment(createdAtMoment, 'day')) : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
   
    return startDateMatch && endDateMatch && textMatch
     
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    }
    if (sortBy === 'amount') {
      return a.amount > b.amount ? -1 : 1;
    }
  });

};