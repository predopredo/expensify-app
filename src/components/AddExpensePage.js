// Npm modules
import React from 'react';
// Actions
import { addExpense } from '../actions/expenses'
// react-redux connect
import { connect } from 'react-redux';
// Components
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
  onSubmit = (formExpense) => { //*
    this.props.addExpense(formExpense); // ***
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmitToDispatch={this.onSubmit} //*
        />
      </div>
    );
  }
};

// mapDispatchToProps let's you can return an object that somehow uses dispatch
// here we use it to abstract it away from component so this component gets easily testable with spy function
const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense)) // ***
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage);