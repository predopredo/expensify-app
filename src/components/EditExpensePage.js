// Npm modules
import React from 'react';
// react-redux
import { connect } from 'react-redux';
// actions
import { editExpense, removeExpense } from '../actions/expenses';
// Components
import ExpenseForm from './ExpenseForm';
import NotFoundPage from '../components/NotFoundPage'

export class EditExpensePage extends React.Component {
  render() {
    return this.props.expense ? (
      <div>
        <h1>Edit Expense</h1>
        <ExpenseForm
          expense={this.props.expense}
          onSubmitToDispatch={(FormExpense) => {
            this.props.editExpense(this.props.expense.id, FormExpense);
            this.props.history.push('/');
          }}
        />
        <button
          onClick={() => {
            this.props.removeExpense(this.props.expense.id);
            this.props.history.push('/');
          }}>Remove</button>
      </div>
    ) :
      <div>
        <NotFoundPage />
      </div>
  }
}
// props = EditExpensesPage props
const mapStateToProps = (storedState, props) => ({
  expense: storedState.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (id) => dispatch(removeExpense({ id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);