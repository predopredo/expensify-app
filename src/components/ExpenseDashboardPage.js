// Npm modules
import React from 'react';
// Components
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;