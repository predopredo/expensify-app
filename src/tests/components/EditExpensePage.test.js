// NPM MODULES
import React from 'react';
import { shallow } from 'enzyme';
// COMPONENTS
import { EditExpensePage } from '../../components/EditExpensePage';
// FIXTURES
import expenses from '../fixtures/expenses';

//*** GLOBALS ***
let editExpenseSpy, removeExpenseSpy, historySpy, wrapper;
const mockExpense = expenses[2];

//*** SETTING UP TESTS ***
beforeEach(() => {
  editExpenseSpy = jest.fn();
  removeExpenseSpy = jest.fn();
  historySpy = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      history={historySpy}
      expense={mockExpense}
      editExpense={editExpenseSpy}
      removeExpense={removeExpenseSpy}
    />
  );
});

//*** RENDERING ***
test('should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render NotFoundPage when no expense matched in url :id', () => {
  wrapper = shallow(<EditExpensePage editExpense={editExpenseSpy} removeExpense={removeExpenseSpy} />);
  expect(wrapper).toMatchSnapshot();
});

//*** FUNCTION CALLING ***
test('should handle editExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmitToDispatch')(mockExpense);
  expect(editExpenseSpy).toHaveBeenLastCalledWith(mockExpense.id, mockExpense);
  expect(historySpy.push).toHaveBeenLastCalledWith('/')
});

test('should handle removeExpense', () => {
  wrapper.find('button').simulate('click')
  expect(removeExpenseSpy).toHaveBeenLastCalledWith(mockExpense.id);
  expect(historySpy.push).toHaveBeenLastCalledWith('/')
});