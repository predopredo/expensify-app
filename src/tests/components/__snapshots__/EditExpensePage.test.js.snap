// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <h1>
    Edit Expense
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
    onSubmitToDispatch={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;

exports[`should render NotFoundPage when no expense matched in url :id 1`] = `
<div>
  <NotFoundPage />
</div>
`;
